### PR TITLE
Custom timestamp for subsegment

### PR DESCRIPTION
**Pull request by srprash**
*Issue #, if available:* [76](https://github.com/aws/aws-xray-sdk-dotnet/issues/76)

*Description of changes:*
Added timestamp parameter to BeginSubsegment and EndSubsegment.
Added Unit tests for the same.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sdk/test/UnitTests/AWSXRayRecorderTests.cs [hunk 1]
```diff
@@ -1052,6 +1052,35 @@ public void TestDefaultStreamingStrategyWithNegativeValue()
             Assert.ThrowsException<ArgumentException>(() => new AWSXRayRecorderBuilder().WithStreamingStrategy(new DefaultStreamingStrategy(-100)));
         }
 
+        [TestMethod]
+        public void TestBeginSubsegmentWithCustomTime()
```
Comment (yogiraj07) [inline sdk/test/UnitTests/AWSXRayRecorderTests.cs]: Sanity check : unit test that checks begin/end subsegment if no value is passed? The assertion should be on the current time.

File: sdk/src/Core/AWSXRayRecorder.netcore.cs [hunk 2]
```diff
@@ -258,19 +259,27 @@ internal void AddFacadeSegment(String name = null)
...
+        private void AddSubsegmentInLambdaContext(string name, DateTime? timestamp = null)
         {
             // If the request is not sampled, the passed subsegment will still be available in TraceContext to
             // stores the information of the trace. The trace information will still propagated to 
             // downstream service, in case downstream may overwrite the sample decision.
             Entity parentEntity = TraceContext.GetEntity();
+            Subsegment subsegment = new Subsegment(name);
             parentEntity.AddSubsegment(subsegment);
             subsegment.Sampled = parentEntity.Sampled;
-            subsegment.SetStartTimeToNow();
```
Comment (yogiraj07) [inline sdk/src/Core/AWSXRayRecorder.netcore.cs]: Make sure unit tests are covered for Lambda workflow.

File: sdk/src/Core/AWSXRayRecorder.netcore.cs [hunk 3]
```diff
@@ -203,12 +204,12 @@ public override void BeginSubsegment(string name)
...
         /// Begin a tracing subsegment. A new subsegment will be created and added as a subsegment to previous facade segment or subsegment.
         /// </summary>
-        private void ProcessSubsegmentInLambdaContext(string name)
+        private void ProcessSubsegmentInLambdaContext(string name, DateTime? timestamp = null)
         {
             if (!TraceContext.IsEntityPresent()) // No facade segment available and first subsegment of a subsegment branch needs to be added
             {
                 AddFacadeSegment(name);
-                AddSubsegmentInLambdaContext(new Subsegment(name));
+                AddSubsegmentInLambdaContext(name, timestamp);
```
Comment (yogiraj07) [inline sdk/src/Core/AWSXRayRecorder.netcore.cs]: Avoid duplicate code -  `AddSubsegmentInLambdaContext(new Subsegment(name,timestamp))` will do necessary setting of the start time. Same thing for `AddSubsegment()`. In the current function definitions you just remove code line that sets the start time.

File: sdk/src/Core/Sampling/TimeStamp.cs [hunk 4]
```diff
@@ -94,7 +94,7 @@ internal bool IsAfter(TimeStamp timeStamp)
 
         private static decimal GetUnixSeconds(DateTime utcNow)
```
Comment (yogiraj07) [inline sdk/src/Core/Sampling/TimeStamp.cs]: Was this a bug?

Comment (srprash) [reply sdk/src/Core/Sampling/TimeStamp.cs]: Yes. It was always returning the current time as unix seconds.
It is a helper method primarily called by SetStartTime(timestamp) and SetEndTime(timestamp)

File: sdk/src/Core/AWSXRayRecorder.netcore.cs (hunk 3 above)
Comment (srprash) [reply sdk/src/Core/AWSXRayRecorder.netcore.cs]: Currently we don't set properties other than `name` for new `Subsegment` instance.

**Pull request merged**